gio/src/subclass/file_monitor.rs

```diff
@@ -0,0 +1,164 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use glib::{prelude::*, subclass::prelude::*, translate::*};
+
+use crate::{ffi, FileMonitor};
+
+pub trait FileMonitorImpl: ObjectImpl + ObjectSubclass<Type: IsA<FileMonitor>> {
+    fn cancel(&self) -> bool {
+        self.parent_cancel()
+    }
+}
+
+pub trait FileMonitorImplExt: FileMonitorImpl {
+    fn parent_cancel(&self) -> bool {
+        unsafe {
+            let data = Self::type_data();
+            let parent_class = data.as_ref().parent_class() as *const ffi::GFileMonitorClass;
+
+            let f = (*parent_class)
+                .cancel
+                .expect("No parent class implementation for \"cancel\"");
+
+            let res = f(self.obj().unsafe_cast_ref::<FileMonitor>().to_glib_none().0);
+            from_glib(res)
+        }
+    }
+}
+
+impl<T: FileMonitorImpl> FileMonitorImplExt for T {}
+
+unsafe impl<T: FileMonitorImpl> IsSubclassable<T> for FileMonitor {
+    fn class_init(class: &mut ::glib::Class<Self>) {
+        Self::parent_class_init::<T>(class);
+
+        let klass = class.as_mut();
+        klass.cancel = Some(cancel::<T>);
+    }
+}
+
+unsafe extern "C" fn cancel<T: FileMonitorImpl>(
+    monitor: *mut ffi::GFileMonitor,
+) -> glib::ffi::gboolean {
+    let instance = &*(monitor as *mut T::Instance);
+    let imp = instance.imp();
+
+    let res = imp.cancel();
+
+    res.into_glib()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::cell::RefCell;
+
+    use super::*;
+    use crate::{prelude::*, File, FileMonitorEvent};
+
+    mod imp {
+        use super::*;
+
+        #[derive(Default)]
+        pub struct MyFileMonitor {
+            pub index: RefCell<i8>,
+        }
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for MyFileMonitor {
+            const NAME: &'static str = "MyFileMonitor";
+            type Type = super::MyFileMonitor;
+            type ParentType = FileMonitor;
+        }
+
+        impl ObjectImpl for MyFileMonitor {}
+
+        impl FileMonitorImpl for MyFileMonitor {
+            fn cancel(&self) -> bool {
+                self.index.replace(-1);
+                true
+            }
+        }
+    }
+
+    glib::wrapper! {
+        pub struct MyFileMonitor(ObjectSubclass<imp::MyFileMonitor>) @extends FileMonitor;
+    }
+
+    impl MyFileMonitor {
+        pub async fn tick(&self) {
+            std::thread::sleep(std::time::Duration::from_millis(10));
```

Comment:
As this is an async function it should probably use an async sleep mechanism. `glib::timeout_future()` for example